docs/developers/architecture.rst

```diff
@@ -0,0 +1,128 @@
+syslog messages. We called them:
+
+- Listener
+- Server
+- A worker per platform
+- Publisher
+
+The processes communicate between them using
+`ZeroMQ IPC <http://api.zeromq.org/2-1:zmq-ipc>`_.
+
+Listener
+++++++++
+
+The Listener has the role to receive the syslog snippets and queue them
+(``PUSH``) into an IPC socket to the Server. This has two advantages: we ensure
+we queue the messages immediately as received - no time wasted triaging (this
+is very important when the messages are received over UDP, for example -- see
+the :ref:`listener-udp` listener). The other gain by doing so is that we don't
+loose any messages: not even when the napalm-logs process crashes or is
+intentionally stopped: after restart, the Server will continue dequeuing
+messages from that buffer.
+
+The Listener is a pluggable interface, check :ref:`listener` for more details.
+
+The communication between the Listener and the Server is a straight PUSH-PULL
+socket.
+
+Server
+++++++
+
+The Server is the process deals with the triage: using the :ref:`device-profiles`,
+it identifies the platform it comes from. Using this information, the messages
+will be queued to the corresponding worker (see next section). The pattern in
+this case is a
+`Ventiallator Sink <http://zguide.zeromq.org/py:all#Divide-and-Conquer>`_,
+more specifically implemented using
+`ROUTER and DEALER <http://zeromq.org/tutorials:dealer-and-router>`_
+sockets, where the Server is the ROUTER, and each Platform Worker is a DEALER.
+
+The messages at this point are partially parsed, and they can be published
+using one of the available :ref:`publisher`, through the :ref:`logger`
+interface.
+When unable to identify the platform, the message is by default discarded.
+However, the user can activate the messages to be published using the
+:ref:`publisher-opts-send-unknown` option, the format being :ref:`UNKNOWN`.
+Note that the Logger interface has a similar option,
+:ref:`logger-opts-send-unknown`.
+
+Platform Worker
++++++++++++++++
+
+There is one worker started per platform. Each worker receives the partially
+processed messages from the Server, then extracts the data and mapps it to the
+OpenConfig or IETF YANG model, as configured in the :ref:`device-profiles`.
+When a message does not have a corresponding profile mapping, it is discarded.
+To receive however these messages, the user can choose to publish them
+using the :ref:`publisher-opts-send-raw` option.
+
+The messages are then sent to the Publisher IPC socket using ``PUSH``.
+
+You can avoid unwanted workers using the :ref:`configuration-options-device-blacklist`
+and :ref:`configuration-options-device-whitelist` options.
+
+Publisher
++++++++++
+
+The Publisher process retrieves the messages from the IPC socket using ``PULL``
+operations, then forward them over the :ref:`publisher` interfaces. When the
+messages encryption is not turned off (see
+:ref:`configuration-options-disable-security`), the Publisher has also the role
```

Comment:
`the Publisher has also the role to encrypt and sign before publishing`

could be:

`the Publisher also has the role of encrypting and signing before publishing`

or

`the Publisher also has to encrypt and sign before publishing`

or

`the Publisher also encrypts and signs before publishing`